File: backend/src/auth/rbac.py
# Role-Based Access Control (RBAC) implementation

from enum import Enum
from functools import wraps
from flask_jwt_extended import get_jwt, get_jwt_identity
from flask import jsonify

class Role(Enum):
    """User roles with hierarchical permissions"""
    DEVELOPER = 'developer'
    TEAM_LEAD = 'team_lead'
    ADMIN = 'admin'


DEVELOPER_PERMISSIONS = [
    'can_view_tasks',
    'can_update_assigned_tasks',
    'can_comment_on_tasks',
    'can_view_notifications',
    'can_manage_personal_notifications',
    'can_view_own_profile',
    'can_update_own_profile',
    'can_link_github_account',
]

TEAM_LEAD_PERMISSIONS = [
    *DEVELOPER_PERMISSIONS,
    'can_create_tasks',
    'can_assign_tasks',
    'can_update_any_task',
    'can_view_all_users',
    'can_view_system_stats',
    'can_generate_reports',
]

ADMIN_PERMISSIONS = [
    *TEAM_LEAD_PERMISSIONS,
    'can_manage_users',
    'can_manage_projects',
    'can_manage_system_settings',
    'can_view_audit_logs',
    'can_link_github_repos',
]


ROLE_PERMISSIONS = {
    Role.DEVELOPER.value: DEVELOPER_PERMISSIONS,
    Role.TEAM_LEAD.value: TEAM_LEAD_PERMISSIONS,
    Role.ADMIN.value: ADMIN_PERMISSIONS,
}


# ---------------------------------------------------------------------------
# Role hierarchy – higher value = more privileged
# ---------------------------------------------------------------------------
ROLE_HIERARCHY = {
    Role.DEVELOPER: 0,
    Role.TEAM_LEAD: 1,
    Role.ADMIN: 2,
}


def _role_level(role_value):
    """Return the numeric hierarchy level for a role string value."""
    for role_enum, level in ROLE_HIERARCHY.items():
        if role_enum.value == role_value:
            return level
    return -1


def has_permission(role_value, permission):
    """Check whether *role_value* (a string such as 'admin') grants *permission*."""
    return permission in ROLE_PERMISSIONS.get(role_value, [])


# ---------------------------------------------------------------------------
# Decorators
# ---------------------------------------------------------------------------

def require_role(role):
    """Decorator to require a specific role"""
    def decorator(fn):
        @wraps(fn)
        def wrapper(*args, **kwargs):
            claims = get_jwt()
            user_role = claims.get('role')
            expected_role = role.value if isinstance(role, Role) else role
            
            if not user_role or user_role != expected_role:
                return jsonify({'message': 'Insufficient role permissions'}), 403
            
            return fn(*args, **kwargs)
        return wrapper
    return decorator


def require_permission(permission):
    """Decorator to require a specific permission"""
    def decorator(fn):
        @wraps(fn)
        def wrapper(*args, **kwargs):
            claims = get_jwt()
            user_role = claims.get('role')
            
            if not user_role or permission not in ROLE_PERMISSIONS.get(user_role, []):
                return jsonify({'message': 'Insufficient permissions'}), 403
            
            return fn(*args, **kwargs)
        return wrapper
    return decorator


def role_at_least(min_role):
    """Decorator that requires the caller's role to be *min_role* or higher.

    Example::

        @role_at_least(Role.TEAM_LEAD)
        def some_view(): ...
    """
    min_role_enum = min_role if isinstance(min_role, Role) else Role(min_role)
    min_level = ROLE_HIERARCHY[min_role_enum]

    def decorator(fn):
        @wraps(fn)
        def wrapper(*args, **kwargs):
            claims = get_jwt()
            user_role = claims.get('role')
            if not user_role or _role_level(user_role) < min_level:
                return jsonify({'message': 'Insufficient permissions'}), 403
            return fn(*args, **kwargs)
        return wrapper
    return decorator

File: frontend/src/utils/rbac.js
// RBAC Constants and Helpers

export const ROLES = {
  DEVELOPER: 'developer',
  TEAM_LEAD: 'team_lead',
  ADMIN: 'admin'
};

export const ROLE_HIERARCHY = {
  [ROLES.DEVELOPER]: 0,
  [ROLES.TEAM_LEAD]: 1,
  [ROLES.ADMIN]: 2
};

// Available permissions matching backend
export const PERMISSIONS = {
  CAN_MANAGE_PROJECTS: 'can_manage_projects',
  CAN_ASSIGN_TASKS: 'can_assign_tasks',
  CAN_UPDATE_ANY_TASK: 'can_update_any_task',
  CAN_VIEW_ALL_USERS: 'can_view_all_users',
  CAN_MANAGE_USERS: 'can_manage_users',
  CAN_MANAGE_SYSTEM_SETTINGS: 'can_manage_system_settings',
  CAN_VIEW_SYSTEM_STATS: 'can_view_system_stats',
  CAN_LINK_GITHUB_ACCOUNT: 'can_link_github_account',
  CAN_LINK_GITHUB_REPOS: 'can_link_github_repos',
  CAN_COMMENT_ON_TASKS: 'can_comment_on_tasks',
  CAN_MANAGE_PERSONAL_NOTIFICATIONS: 'can_manage_personal_notifications'
};

// We don't redefine the map here, the backend controls the actual map and we fetch it.
// These helpers just act on the user's role/permissions list.

export const hasRole = (userRole, targetRole) => {
  return userRole === targetRole;
};

export const hasAnyRole = (userRole, targetRoles) => {
  return targetRoles.includes(userRole);
};

export const roleAtLeast = (userRole, minRole) => {
  const userLevel = ROLE_HIERARCHY[userRole] ?? -1;
  const minLevel = ROLE_HIERARCHY[minRole] ?? 0;
  return userLevel >= minLevel;
};

export const hasPermission = (userPermissions, targetPermission) => {
  return Array.isArray(userPermissions) && userPermissions.includes(targetPermission);
};
